perceval/backends/core/zulip.py

```diff
@@ -0,0 +1,131 @@
+# -*- coding: utf-8 -*-
+#
+# Copyright (C) 2015-2020 Bitergia
+#
+# This program is free software; you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation; either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program. If not, see <http://www.gnu.org/licenses/>.
+#
+# Authors:
+#     Venu Vardhan Reddy Tekula <venuvardhanreddytekula8@gmail.com>
+#
+
+
+import logging
+
+from grimoirelab_toolkit.uris import urijoin
+
+from ...backend import (Backend,
+                        BackendCommand,
+                        BackendCommandArgumentParser)
+from ...client import HttpClient
+
+logger = logging.getLogger(__name__)
+
+CATEGORY_MESSAGE = "message"
+
+
+class Zulip(Backend):
+    """Zulip Backend."""
+
+    version = '0.1.0'
+
+    CATEGORIES = [CATEGORY_MESSAGE]
+
+    def __init__(self, url, stream, api_token, email, origin, tag=None, archive=None, ssl_verify=True):
+        super().__init__(origin, tag, archive, ssl_verify)
+        origin = url
+
+        super.__init__(origin, tag=tag, archive=archive, ssl_verify=True)
+        self.stream = stream
+        self.api_token = api_token
+        self.email = email
+
+        self.client = None
+
+    def fetch(self, category=CATEGORY_MESSAGE):
+        """Fetch the message from the stream."""
+        kwargs = {}
+        items = super().fetch(category, **kwargs)
+
+        return items
+
+    def fetch_items(self, category, **kwargs):
+        """Fetch the messages."""
+
+        anchor = 'oldest'
+        raw_message = self.client.get_message(anchor)
+        print(raw_message)
+
+    def _init_client(self, from_archive=False):
+        """Init client"""
+
+        return ZulipClient(self.url, self.stream, self.api_token,
+                           self.email, from_archive, self.ssl_verify)
+
+
+class ZulipClient(HttpClient):
+    """Zulip API client."""
+
+    def __init__(self, url, stream, api_token, email,
+                 archive=None, from_archive=False, ssl_verify=True):
+        self.stream = stream
+        self.api_token = api_token
+        self.email = email
+
+        super().__init__(url,
+                         archive=archive, from_archive=from_archive, ssl_verify=ssl_verify)
+
+    def fetch(self, url, payload=None, headers=None):
+        response = super().fetch(url, payload)
+        return response
+
+    def get_message(self, anchor):
+        request = {
```

Comment:
To get data from this API we need to pass the token and email. Looking at the [zulip wrapper](https://github.com/zulip/python-zulip-api/blob/16737e4ed3ecd9dc6b1b25f4566f3477ed96b089/zulip/zulip/__init__.py#L419), they rely on HTTPBasicAuth. We can leverage on the param `auth` provided by the generic Perceval fetch method.

Suggested change:
```
+        params = {
             'anchor': '{}'.format(anchor),
             'num_before': '0',
             'num_after': '2',
             'apply_markdown': 'false',
             'narrow': json.dumps([{'operator': 'stream', 'operand': '{}'.format(self.stream)}])
         }
-        path = urijoin(url, "/api/v1/messages")
-        payload = {'params': request}
+        path = urijoin(self.url, "/api/v1/messages")
+        r = self.fetch(path, payload=params, auth=(self.email, self.api_token))
 
-        return self.fetch(path, payload)
+        return r.text
```

Reply:
Thanks for the suggestion @valeriocos.

I tried this, ended up with an error
```
RuntimeError: fetch() got an unexpected keyword argument 'auth'
```

I think it is because the argument `auth`. I resolved it with this
```
-    def fetch(self, url, payload=None, headers=None):
+    def fetch(self, url, payload=None, headers=None, auth=None):
```
The previous error was not persistent, but got a new one
```
RuntimeError: 401 Client Error: Unauthorized for url: https://python.zulipchat.com/api/v1/messages?anchor=oldest&num_before=0&num_after=2&apply_markdown=false&narrow=%5B%7B%22operator%22%3A+%22stream%22%2C+%22operand%22%3A+%22importlib%22%7D%5D
```

I was using the working token and email. I regenerated and tested with this [script](https://gist.github.com/vchrombie/bfd6d967e03bbe28cf4241995fb4b91b#file-script-py).

Did you have the same error?

Reply:
I understand that the argument `auth` should be passed to these functions, [client.py#L143 ](https://github.com/chaoss/grimoirelab-perceval/blob/master/perceval/client.py#L143 ) & [client.py#L174](https://github.com/chaoss/grimoirelab-perceval/blob/master/perceval/client.py#L174). Let me know if I am wrong.

But I see a `None` object passing through it. :thinking:  :confused:

Reply:
The method fetch at https://github.com/chaoss/grimoirelab-perceval/pull/667/files#diff-55fae40585a834540929c7f91a9c1e89R194, it isn't needed. The call at https://github.com/chaoss/grimoirelab-perceval/pull/667/files#diff-55fae40585a834540929c7f91a9c1e89R220 will use the parent `fetch`

Reply:
Sorry for the confusion, I got it right now. :+1:

Reply:
No worries, sorry for the mistake with the diffs in the first iteration